const-generic-const-fn-bounds.md

```diff
@@ -0,0 +1,279 @@
+- Feature Name: const_generic_const_fn_bounds
+- Start Date: 2018-10-05
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow `impl const Trait` for trait impls where all method impls are checked as const fn.
+
+Make it legal to declare trait bounds on generic parameters of const functions and allow
+the body of the const fn to call methods on the generic parameters that have a `const` modifier
+on their bound.
+
+# Motivation
+[motivation]: #motivation
+
+Currently one can declare const fns with generic parameters, but one cannot add trait bounds to these
+generic parameters. Thus one is not able to call methods on the generic parameters (or on objects of the
+generic parameter type), because they are fully unconstrained.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+You can call methods of generic parameters of a const function, because they are implicitly assumed to be
+`const fn`. For example, the `Add` trait declaration has an additional `const` before the trait name, so
+you can use it as a trait bound on your generic parameters:
+
+```rust
+const fn triple_add<T: const Add>(a: T, b: T, c: T) -> T {
+    a + b + c
+}
+```
+
+The obligation is passed to the caller of your `triple_add` function to supply a type whose `Add` impl is fully
+`const`. Since `Add` only has `add` as a method, in this case one only needs to ensure that the `add` method is
+`const`. Instead of adding a `const` modifier to all methods of a trait impl, the modifier is added to the entire
+`impl` block:
+
+```rust
+struct MyInt(i8);
+impl const Add for MyInt {
+    fn add(self, other: Self) -> Self {
+        MyInt(self.0 + other.0)
+    }
+}
+```
+
+The const requirement is required on all bounds of the impl and its methods,
+so in the following `H` is required to have a const impl of `Hasher`, so that
+methods on `state` are callable.
+
+```rust
+impl const Hash for MyInt {
+    const fn hash<H>(
+        &self,
+        state: &mut H,
+    )
+        where H: const Hasher
+    {
+        state.write(&[self.0 as u8]);
+    }
+}
+```
+
+While these `const` keywords could be inferred (after all, they are required), requiring them is
+forward compatible to schemes in the future that allow more fine grained control.
+
+## Drop
+
+A notable use case of `impl const` is defining `Drop` impls. If you write
+
+```rust
+struct SomeDropType<'a>(&'a Cell<u32>);
+impl const Drop for SomeDropType {
+    fn drop(&mut self) {
+        self.0.set(self.0.get() - 1);
+    }
+}
+```
+
+Then you are allowed to actually let a value of `SomeDropType` get dropped within a constant
```

Comment:
Does the `const Drop` impl also require then that `SomeDropType`'s fields are all `const`-droppable as well?

Reply:
(in contrast to just checking that `fn drop` is const-compatible)

Reply:
Yes, I overlooked that. They must be, as drop glue would be invoking their `Drop` impls, too.

Reply:
Presumably types which don't implement `Drop` manually but have fields that do automatically inherit `const`-`Drop`-ness from their fields? How would one specify a `const`-`Drop`-glue bound without requiring types to implement the actual `Drop` trait?

Reply:
hm. good question. that is an issue. We can't infer it, because that would break code when you change a private field from `const Drop` to just `Drop`. I do not have a good idea for this situation. Suggestions are very welcome.

Reply:
It's gross, but maybe `const drop` for things that have const-compatible `drop` glue rather than `const Drop` for things that `const`-ly implement the `Drop` trait? That's confusing as heck, sorry, but I don't have other good ideas.